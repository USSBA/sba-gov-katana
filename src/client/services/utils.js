import querystring from 'querystring'

function getQueryParams() {
  let queryParams = {}
  const { search } = window.location
  if (search && search !== '') {
    queryParams = querystring.decode(search.replace('?', ''))
  }
  return queryParams
}

export { getQueryParams }
