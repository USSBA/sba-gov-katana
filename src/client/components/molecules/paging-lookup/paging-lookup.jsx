import React from "react";
import {connect} from "react-redux";
import {bindActionCreators} from "redux";
import _ from "lodash";

import styles from "./paging-lookup.scss";

import * as ContentActions from "../../../actions/content.js";
import {logEvent} from "../../../services/analytics.js";
import {getQueryParams} from "../../../services/utils.js";
import {DocumentArticleLookup} from "molecules";

const config = {
  pageSize: 30
};

class PagingLookup extends React.Component {

  constructor(ownProps) {
    super();
    this.state = this.createOriginalState(ownProps);
  }

  createOriginalState(ownProps) {
    return {query: this.createQueryFromProps(ownProps), taxonomies: [], pageNumber: 1, isFetching: false};
  }

  createQueryFromProps(ownProps) {
    let propsSource = ownProps;
    let defaults = _.chain(propsSource.taxonomyFilters).keyBy().mapValues(_ => "All").value();

    let queryParams = getQueryParams();

    // look for aliases in the query params, but filter out any that are undefined
    let aliasMapping = _.filter({
      searchTerm: queryParams.search || queryParams.q,
      documentType: queryParams.type,
      documentActivity: queryParams.activity
    }, value => value)

    let filteredQueryParams = _.filter(queryParams, (value, key) => {
      return _.includes(propsSource.taxonomyFilters, key)
    })

    let finalQuery = _.assign({
      sortBy: "Last Updated",
      searchTerm: ""
    }, defaults, queryParams, aliasMapping)
    return finalQuery;
  }

  componentWillMount() {
    this.props.actions.fetchContentIfNeeded("taxonomies", "taxonomyVocabulary", {"names": this.props.taxonomyFilters.join(",")});
  }

  componentWillReceiveProps(nextProps, ownProps) {
    const updatedProps = {};

    // set the taxomonies object when
    // nextProps.items array is populated, AND state.taxonomies array is NOT populated

    if (nextProps.taxonomies && nextProps.taxonomies.length > 0 && this.state.taxonomies.length === 0) {

      const taxonomies = nextProps.taxonomies.slice();

      // add an "All" filter option to dynamic taxonomies
      for (let index = 0; index < taxonomies.length; index++) {
        taxonomies[index].terms.unshift("All");
      }

      const rearrangedTaxonomyOrder = _.map(this.props.taxonomyFilters, (taxonomyVocabularyName) => {
        return _.find(taxonomies, {name: taxonomyVocabularyName})
      })

      // add a "Sort By" taxonomy object to append a "Sort By" multiselect component
      rearrangedTaxonomyOrder.push({
        "name": "Sort By",
        "terms": this.props.sortByOptions
      });

      updatedProps.taxonomies = rearrangedTaxonomyOrder;
    }

    if (nextProps.itemReponse) {
      updatedProps.isFetching = false;
    }

    this.setState(updatedProps);

  }

  handleReset() {
    this.setState(_.assign(this.createOriginalState(this.props),{taxonomies: this.state.taxonomies}), ()=>{
        this.submit();
    });
  }

  componentDidMount() {
    this.submit();
  }

  handleSubmit() {
    this.setState({
      items: undefined,
      pageNumber: 1
    }, () => {
      this.submit();
    });
  }

  submit() {
    this.setState({
      isFetching: true
    }, () => {
      const start = ((this.state.pageNumber - 1) * config.pageSize);
      const end = start + config.pageSize;

      const remappedState = _.mapValues(this.state.query, (value) => {
        return value === "All"
          ? "all"
          : value;
      })
      const queryTerms = _.assign({}, remappedState, {start, end});

      this.props.actions.fetchContentIfNeeded(this.props.type, this.props.type, queryTerms);

      this.fireEvent("document-lookup", "Filter Status : " + JSON.stringify(this.state.query), null);

    });
  }

  fireEvent(category, action, value) {
    logEvent({category: category, action: action, label: window.location.pathname, value: value})
  }

  handleQueryChange(field, value) {
    let newQueryFieldValue = {};
    newQueryFieldValue[field] = value;
    let currentQuery = this.state.query;
    let newQuery = _.assign({}, currentQuery, newQueryFieldValue);
    this.setState({query: newQuery});
  }

  handlePageChange(newPageNumber) {
    this.setState({
      pageNumber: newPageNumber
    }, _ => this.submit());
  }

  render() {
    let lookupProps = {
      title: this.props.title,
      queryState: this.state.query,
      items: this.props.itemReponse.items,
      itemCount: this.props.itemReponse.count,
      pageNumber: this.state.pageNumber,
      pageSize: config.pageSize,
      taxonomies: this.state.taxonomies,
      onSubmit: this.handleSubmit.bind(this),
      onReset: this.handleReset.bind(this),
      onQueryChange: this.handleQueryChange.bind(this),
      onPageChange: this.handlePageChange.bind(this),
      isFetching: this.state.isFetching,
      fieldsToShowInDetails: this.props.fieldsToShowInDetails,
      type: this.props.type
    }
    return (<DocumentArticleLookup {...lookupProps}/>)

  }
}

PagingLookup.propTypes = {
  title: React.PropTypes.string,
  type: React.PropTypes.string,
  taxonomyFilters: React.PropTypes.array,
  fieldsToShowInDetails: React.PropTypes.array
}

PagingLookup.defaultProps = {
  title: "",
  type: "",
  taxonomyFilters: [],
  fieldsToShowInDetails: [],
  itemReponse: {
    items: undefined,
    count: 0
  }
}

function mapReduxStateToProps(reduxState, ownProps) {
  const {taxonomies} = reduxState.contentReducer;
  return {
    itemReponse: reduxState.contentReducer[ownProps.type],
    taxonomies: taxonomies
  };
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators(ContentActions, dispatch)
  }
}
export default connect(mapReduxStateToProps, mapDispatchToProps)(PagingLookup);
