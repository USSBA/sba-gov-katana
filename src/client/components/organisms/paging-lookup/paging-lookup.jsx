import React from 'react'
import PropTypes from 'prop-types'
import { browserHistory } from 'react-router'
import { assign, chain, find, includes, map, mapValues, pickBy } from 'lodash'
import { DocumentArticleLookup } from 'organisms'
import styles from './paging-lookup.scss'
import { getQueryParams } from '../../../services/utils.js'
import { fetchSiteContent } from '../../../fetch-content-helper'
import { logEvent } from '../../../services/analytics.js'

const config = {
  pageSize: 30
}

const fieldNameMap = {
  searchTerm: 'Search',
  documentType: 'Type',
  program: 'Program',
  documentActivity: 'Activity',
  office: 'Office',
  sortBy: 'sortBy'
}

class PagingLookup extends React.Component {
  constructor(ownProps) {
    super()
    this.state = this.createOriginalState(ownProps)
  }

  createOriginalState(ownProps) {
    return {
      query: this.createQueryParams(ownProps),
      taxonomies: [],
      pageNumber: 1,
      isFetching: false,
      itemResponse: {
        items: null,
        count: 0
      }
    }
  }

  createQueryParams(ownProps) {
    const query = this.createQueryFromProps(ownProps)
    query.office = query.office || 'All'
    return query
  }

  createQueryFromProps(ownProps) {
    const propsSource = ownProps
    const defaults = chain(propsSource.taxonomyFilters)
      .keyBy()
      .mapValues(_ => {
        return 'All'
      })
      .value()

    const queryParams = getQueryParams()
    // look for aliases in the query params, but filter out any that are undefined
    const aliasMapping = pickBy({
      searchTerm: queryParams.search || queryParams.q,
      documentType: queryParams.type,
      documentActivity: queryParams.activity,
      office: queryParams.office,
      page: Number(queryParams.page) || 1
    })

    const filteredQueryParams = pickBy(queryParams, (value, key) => {
      return includes(propsSource.taxonomyFilters, key)
    })

    const finalQuery = assign(
      {
        sortBy: queryParams.sortBy || ownProps.defaultSortBy,
        searchTerm: ''
      },
      defaults,
      filteredQueryParams,
      aliasMapping
    )
    return finalQuery
  }

  handleReset(ownProps) {
    this.fireDocumentationLookupEvent('clearFilters')
    this.setState(
      assign(this.createOriginalState(this.props), {
        query: {
          searchTerm: '',
          documentActivity: 'All',
          documentType: 'All',
          program: 'All',
          office: 'All',
          sortBy: this.props.defaultSortBy
        },
        taxonomies: this.state.taxonomies,
        pageNumber: 1,
        isFetching: false
      }),
      () => {
        this.handleSubmit()
      }
    )
  }

  async componentDidMount() {
    const queryArgs = {
      names: this.props.taxonomyFilters.join(',')
    }

    const taxonomies = await fetchSiteContent('taxonomys', queryArgs)

    this.setState({
      taxonomies: this.reArrangeTaxonomyOrder(taxonomies)
    })
  }

  componentWillReceiveProps({ sbaOffices }) {
    this.setState(
      {
        sbaOfficeFilters: this.createSbaOfficeFilterList(sbaOffices)
      },
      () => this.submit()
    )
  }

  reArrangeTaxonomyOrder(data) {
    const taxonomies = data.slice()

    // add an "All" filter option to dynamic taxonomies
    for (let index = 0; index < taxonomies.length; index++) {
      taxonomies[index].terms.unshift('All')
    }

    const rearrangedTaxonomyOrder = map(this.props.taxonomyFilters, taxonomyVocabularyName => {
      return find(taxonomies, { name: taxonomyVocabularyName })
    })

    // add a "Sort By" taxonomy object to append a "Sort By" multiselect component
    rearrangedTaxonomyOrder.push({
      name: 'Sort By',
      terms: this.props.sortByOptions
    })
    return rearrangedTaxonomyOrder
  }

  createSbaOfficeFilterList(sbaOffices) {
    if (Array.isArray(sbaOffices) === true && sbaOffices.length > 0) {
      const filteredOffices = [
        {
          label: 'All',
          value: 'All'
        }
      ]
      sbaOffices.forEach(office => {
        const filterFormattedOffice = {}
        filterFormattedOffice.label = office.title
        filterFormattedOffice.value = office.id
        filteredOffices.push(filterFormattedOffice)
      })
      return filteredOffices.sort(function(a, b) {
        // Sort alphabetically and push "All" to the top of the list of options
        if (a.label < b.label || a.label === 'All') {
          return -1
        }
        if (a.label > b.label) {
          return 1
        }
        return 0
      })
    }

    return null
  }

  handleSubmit() {
    const queryObject = this.state.query

    this.fireDocumentationLookupEvent(`Apply CTA: Term: ${queryObject.searchTerm}`)
    this.setState(
      {
        items: null,
        pageNumber: 1
      },
      () => {
        this.submit()
      }
    )
    this.handlePageChange(1)
  }

  convertAndSetQueryObjectToString(queryObject) {
    const queryString = this.convertQueryObjectToString(queryObject)
    this.setQueryStringToUrl(queryString)
  }

  convertQueryObjectToString(queryObject) {
    let queryString = ''

    for (const key in queryObject) {
      if (key === 'searchTerm') {
        queryString += `search=${queryObject[key]}&`
      } else {
        queryString += `${key}=${queryObject[key]}&`
      }
    }
    return queryString
  }

  setQueryStringToUrl(queryString) {
    const type = this.props.type

    browserHistory.push({
      pathname: `/${type.slice(0, -1)}`,
      search: `?${queryString.slice(0, -1)}`
    })
  }

  submit() {
    const start = (this.state.query.page - 1) * config.pageSize
    const end = start + config.pageSize

    const remappedState = mapValues(this.state.query, value => {
      return value === 'All' ? 'all' : value
    })
    const queryTerms = assign({}, remappedState, { start, end })

    this.setState(
      {
        isFetching: true
      },
      async () => {
        this.setState({
          itemResponse: await fetchSiteContent(this.props.type, queryTerms),
          isFetching: false
        })
      }
    )
  }

  fireEvent(category, action, value) {
    logEvent({
      category: category,
      action: action,
      label: window.location.pathname,
      value: value
    })
  }

  fireDocumentationLookupEvent(action, value = null) {
    this.fireEvent('documentation-lookup', action, value)
  }

  handleQueryChange(field, value) {
    if (field !== 'searchTerm') {
      // Log Analytic Event, but not for search term
      this.fireDocumentationLookupEvent(`${fieldNameMap[field] || field}: ${value}`)
    }
    const newQueryFieldValue = {}
    newQueryFieldValue[field] = value
    const currentQuery = this.state.query
    const newQuery = assign({}, currentQuery, newQueryFieldValue)
    this.setState({ query: newQuery })
  }

  updatePageNumberinQueryObject(newPageNumber) {
    const queryObject = this.state.query
    const queryObjectWithUpdatedPageNumber = Object.assign(queryObject, {
      page: newPageNumber
    })
    this.convertAndSetQueryObjectToString(queryObjectWithUpdatedPageNumber)
  }

  handlePageChange(newPageNumber) {
    this.updatePageNumberinQueryObject(newPageNumber)
    return this.submit()
  }

  render() {
    const lookupProps = {
      title: this.props.title,
      queryState: this.state.query,
      items: this.state.itemResponse.items,
      itemCount: this.state.itemResponse.count,
      pageNumber: this.state.query.page,
      pageSize: config.pageSize,
      taxonomies: this.state.taxonomies,
      sbaOffices: this.state.sbaOfficeFilters,
      onSubmit: this.handleSubmit.bind(this),
      onReset: this.handleReset.bind(this),
      onQueryChange: this.handleQueryChange.bind(this),
      onPageChange: this.handlePageChange.bind(this),
      isFetching: this.state.isFetching,
      fieldsToShowInDetails: this.props.fieldsToShowInDetails,
      type: this.props.type
    }
    return <DocumentArticleLookup {...lookupProps} />
  }
}

PagingLookup.propTypes = {
  title: PropTypes.string,
  type: PropTypes.string,
  taxonomyFilters: PropTypes.array,
  fieldsToShowInDetails: PropTypes.array,
  sortByOptions: PropTypes.array
}

PagingLookup.defaultProps = {
  title: '',
  type: '',
  taxonomyFilters: [],
  sbaOffices: null,
  fieldsToShowInDetails: [],
  sortByOptions: []
}

export default PagingLookup
