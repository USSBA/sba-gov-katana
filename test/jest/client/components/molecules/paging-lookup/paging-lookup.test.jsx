/* eslint-disable no-undef */
/*global expect*/

import React from 'react'
import { PagingLookup } from 'client/components/molecules/paging-lookup/paging-lookup.jsx'
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer'
import _ from 'lodash'

jest.mock('../../../../../../src/client/services/utils.js')
getQueryParams.mockImplementation(() => {
  return {}
})
import { getQueryParams } from '../../../../../../src/client/services/utils.js'
const fetchContentIfNeeded = jest.fn()

const testProps = {
  title: 'My Documents',
  type: 'documents',
  taxonomyFilters: ['documentType', 'documentActivity', 'validTaxonomyFilter'],
  fieldsToShowInDetails: ['Type', 'Name', 'Summary'],
  sortByOptions: ['Last Updated', 'Size'],
  actions: {
    fetchContentIfNeeded: fetchContentIfNeeded
  }
}

describe('PagingLookup', () => {
  test('should render with all the information', () => {
    const props = _.clone(testProps)
    const component = renderer.create(<PagingLookup {...props} />)
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('should have an empty default query search', () => {
    getQueryParams.mockImplementationOnce(() => {
      return {}
    })
    const component = shallow(<PagingLookup {...testProps} />)
    expect(component.first().props().queryState.searchTerm).toBe('')
    expect(component.first().props().queryState.documentType).toBe('All')
    expect(component.first().props().queryState.documentActivity).toBe('All')
    expect(component.first().props().queryState.validTaxonomyFilter).toBe('All')
  })
  test("should accept a 'search' term", () => {
    getQueryParams.mockImplementationOnce(() => {
      return { search: 'My Search Term' }
    })
    const component = shallow(<PagingLookup {...testProps} />)
    expect(component.first().props().queryState.searchTerm).toBe(
      'My Search Term'
    )
  })
  test("should accept a 'q' term", () => {
    getQueryParams.mockImplementationOnce(() => {
      return { q: 'My Q Search Term' }
    })
    const component = shallow(<PagingLookup {...testProps} />)
    expect(component.first().props().queryState.searchTerm).toBe(
      'My Q Search Term'
    )
  })
  test("should accept 'type' from the query string, converting it to documentType", () => {
    getQueryParams.mockImplementationOnce(() => {
      return { type: 'SBA Form' }
    })
    const component = shallow(<PagingLookup {...testProps} />)
    expect(component.first().props().queryState.documentType).toBe('SBA Form')
  })
  test("should accept 'activity' from the query string, converting it to documentActivity", () => {
    getQueryParams.mockImplementationOnce(() => {
      return { activity: 'All' }
    })
    const component = shallow(<PagingLookup {...testProps} />)
    expect(component.first().props().queryState.documentActivity).toBe('All')
  })
  test("should accept 'validTaxonomyFilter' from the query string if in the taxonomyFilter", () => {
    getQueryParams.mockImplementationOnce(() => {
      return { validTaxonomyFilter: 'My Taxonomy Filter' }
    })
    const component = shallow(<PagingLookup {...testProps} />)
    expect(component.first().props().queryState.validTaxonomyFilter).toBe(
      'My Taxonomy Filter'
    )
  })
  test("should reject 'arbitraryNonsense' from the query string if NOT in the taxonomyFilter", () => {
    getQueryParams.mockImplementationOnce(() => {
      return { arbitraryNonsense: 'My Arbitrary Nonsense' }
    })
    const component = shallow(<PagingLookup {...testProps} />)
    expect(
      component.first().props().queryState.arbitraryNonsense
    ).toBeUndefined()
  })

  // test('should render without an address ', () => {
  //   let cardData = _.omit(card, "streetAddress");
  //   const component = shallow(<LinkCard {...cardData}/>);
  //   expect(component.find(".paging-lookup-streetAddress")).toHaveLength(0);
  // });
  //
  // test('should render without a zip code ', () => {
  //   let cardData = _.omit(card, "zipCode");
  //   const component = shallow(<LinkCard {...cardData}/>);
  //   expect(component.find(".paging-lookup-citystatezip")).toHaveLength(0);
  // });
  //
  // test('should render without the link', () => {
  //   let cardData = _.omit(card, "link");
  //   const component = shallow(<LinkCard {...cardData}/>);
  //   expect(component.find(".paging-lookup-link")).toHaveLength(0);
  // });
})
