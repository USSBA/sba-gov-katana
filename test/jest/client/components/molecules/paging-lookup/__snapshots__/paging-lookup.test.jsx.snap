// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PagingLookup should render with all the information 1`] = `
<div>
  <div
    className="banner"
  >
    <h2
      className="header"
    >
      My Documents
    </h2>
  </div>
  <div
    className="result"
  >
    <div />
    <div
      className="documentCardCollection"
    >
      <div
        className="loaderContainer"
      >
        <div
          className="loader container undefined"
        >
          <img
            className="logo"
            src="test-file-stub"
          />
          <div
            className="loader"
          />
        </div>
      </div>
    </div>
    <div />
  </div>
</div>
`;
